tools/release/build.sh

```diff
@@ -0,0 +1,62 @@
+#!/bin/zsh
+
+# Usage:
+# $ ./tools/release/build.sh -h
+# Builds artifacts for specified tag.
+
+# Options:
+#   --tag: The tag to release.
+#   --artifacts-path: Path to store artifacts.
+
+set -eo pipefail
+source ./tools/utils/argparse.sh
+source ./tools/utils/echo-color.sh
+
+set_description "Builds artifacts for specified tag."
+define_arg "tag" "" "The tag to release." "string" "true" 
+define_arg "artifacts-path" "" "Path to store artifacts." "string" "true"
+
+check_for_help "$@"
+parse_args "$@"
+
+ARTIFACTS_PATH="$artifacts_path"
+REPO_CLONE_PATH="$ARTIFACTS_PATH/dd-sdk-ios"
+XCF_DIR_NAME="Datadog.xcframework"
+XCF_ZIP_NAME="Datadog.xcframework.zip"
+
+# Clone a fresh version of the repo to the artifacts path to ensure that release tools
+# operate on a clean version of the repo, unaltered by any configuration changes.
+clone_repo () {
+    echo_subtitle "Clone repo for '$tag' into '$ARTIFACTS_PATH'"
+    git clone --depth 1 --branch $tag --single-branch git@github.com:DataDog/dd-sdk-ios.git $REPO_CLONE_PATH
+}
+
+# Create XCFrameworks for the repo clone using release tools from the current repo.
+create_xcframeworks () {
+    echo_subtitle "Create '$XCF_DIR_NAME' in '$ARTIFACTS_PATH'"
+    ./tools/release/build-xcframeworks.sh --repo-path "$REPO_CLONE_PATH" \
+        --ios --tvos \
+        --output-path "$ARTIFACTS_PATH/$XCF_DIR_NAME"
+
+    echo_info "Contents of '$ARTIFACTS_PATH/$XCF_DIR_NAME':"
+    ls $ARTIFACTS_PATH/$XCF_DIR_NAME
+}
+
+create_zip_archive () {
+    echo_subtitle "Create '$XCF_ZIP_NAME' in '$ARTIFACTS_PATH'"
+    cd "$ARTIFACTS_PATH" && zip -r -q $XCF_ZIP_NAME $XCF_DIR_NAME
+    cd -
+}
+
+rm -rf "$ARTIFACTS_PATH"
+mkdir -p "$ARTIFACTS_PATH"
+
+clone_repo
```

Comment:
Could we clean the repo state rather cloning a new copy?

Essentially `git clean -fxd`

Reply:
Yes, we can, but this would significantly change the existing release model introduced to Bitrise in [this PR](https://github.com/DataDog/dd-sdk-ios/pull/636). While migrating things to GitLab I wanted to keep it simple, not exposed to new challenges.

In [#636](https://github.com/DataDog/dd-sdk-ios/pull/636), we switched from using "cleanup" scripts to performing a fresh clone of the repo for two reasons:
1. Overlooking some dirty state during "cleanup" led to shipping unnecessary content in three release artifacts.
2. There was no CI-automated way to re-build artifacts for past releases using the new tooling.

While `git clean -fxd` is definitely better than the [custom "cleanup"](https://github.com/DataDog/dd-sdk-ios/blob/920464537fd09d0b7166ff030fae9040138aab01/tools/distribution/make_distro_builds.sh#L26-L42) we used to have before [#636](https://github.com/DataDog/dd-sdk-ios/pull/636), performing a release from the current repo doesn't offer the same level of flexibility in CI. Specifically:

- With the current model, it is possible to test the entire release pipeline on any branch by targeting an existing tag (using `RELEASE_GIT_TAG` and `RELEASE_DRY_RUN` options). I leveraged this several times to test this PR.
- In rare situations, we had to modify release tooling and re-build artifacts for existing tags using a new version of Xcode on CI (possible by using `RELEASE_GIT_TAG=<past release>`, `RELEASE_DRY_RUN=0`, and `OVERWRITE_EXISTING=1`).

Likely, we don't need such a level of CI flexibility these days ☝️, as module stability [was finally fixed in `2.x`](https://datadoghq.atlassian.net/wiki/x/KQFfiw).

However, cleaning the current repo for developing release tooling **in local** might be very cumbersome and impractical, given that we track a lot of dev configuration through `.local.xcconfig` files (which would be wiped out). Performing a fresh clone of the repo into `artifacts/<tag>/` is more convenient and less invasive.

Let me know how does it sound @ganeshnj

Reply:
> but I don't think it is a good approach to have something released other than what the current branch is triggered from

This is definitely not the case here. For `tag` triggers, the flow is this:
- GitLab clones the repo to `~/dd-sdk-ios` for given `GIT_TAG`
- `~/dd-sdk-ios/tools/release/build.sh` starts
- it clones the repo for **the same** `GIT_TAG` into `~/dd-sdk-ios/artifacts/<GIT_TAG>/dd-sdk-ios`
- finally, `~/dd-sdk-ios/tools/release/` automation:
   - builds and publishes XCFramework for `~/dd-sdk-ios/artifacts/<GIT_TAG>/dd-sdk-ios`
   - publishes Cocoapod podspecs for `~/dd-sdk-ios/artifacts/<GIT_TAG>/dd-sdk-ios`

Ultimately, the automation will release exactly the hash that it was triggered for, so we can't speak of anti-patterns.

Then, I agree that this implementation detail might look custom, but it was made intentionally to address concrete problems we hit when releasing SDK that supports several iOS dependency managers, each operating in different model.

> Let's not block the PR 

👍 Also note that all `tools/release/` scripts introduced in this PR do support path parameters making the release of `.` instead of `artifacts/<GIT_TAG>/dd-sdk-ios` very simple to achieve.

Reply:
Thanks for clarification.